### PR TITLE
Replace yarn 1 with pnpm

### PR DESCRIPTION
**Pull request by anomiex**
#### Changes proposed in this Pull Request:

Replace yarn 1 with pnpm.

A few side effects:

* Update a bunch of indirect dependencies.
* Add some missing dependencies.
* Move the js-test-runner into its own "project", like the Jetpack CLI, so we can remove its deps from the monorepo-root package.json.
* Deprecated a bunch of scripts in the monorepo-root package.json.

#### Jetpack product discussion

p9dueE-2R4-p2

#### Does this pull request change what data or activity we track or use?

No

#### Testing instructions:

See [docs/yarn-upgrade.md](https://github.com/Automattic/jetpack/blob/feature/replace-yarn-1-with-pnpm/docs/yarn-upgrade.md) before you start, you'll probably want to do some cleanup before checking this out.

* Try everything, see if it still works.

### CODE REVIEW

**Comment by anomiex** [comment]
* Tests / JS tests is currently failing due to https://github.com/pnpm/pnpm/issues/1768, which should be fixed by https://github.com/pnpm/cmd-shim/pull/36. That PR is already approved upstream, hopefully they merge it and do a release sooner rather than later.
* Tests / Code coverage looks to be failing for the same reason.
* E2E Tests / E2E connection tests is failing on other PRs with the same error, so probably nothing here.

**Comment by kraftbj** [comment]
I'm aiming to review today. Marking as blocked since we'll need communications in place before merge to alert folks of required tooling change.

**Comment by jeherve** [comment]
At this point of the release cycle, would it make sense to wait until after the code freeze / branch off `master` to merge this, to avoid any potential issues that may come up? Merging this early in the release cycle would leave us a full cycle to iron things out.

**Comment by anomiex** [comment]
I think the necessary communications will wind up requiring we hold off until the next release cycle anyway. I'm hoping we don't wind up having to wait longer, actually, for things like p1621520158124600-slack-CBG1CP4EN.

File: docs/development-environment-cli-support.md [hunk 1]
```diff
@@ -12,15 +12,13 @@ tools/check-development-environment.sh
...
-yarn install
+pnpm cli-setup
 ```
+from the monorepo root.
 
-In some instances, this will not work depending on permissions, etc. You may need to add `~/.yarn/bin` to your system's PATH. 
-If `jetpack` is still not available, you can run `yarn jetpack` from the repo root only.
-
-`jetpack`, when installed, will work from anywhere within the repo.
+You can also use `pnpx --no jetpack` from (almost) anywhere in the monorepo. If you commonly use multiple checkouts, you might add this as a shell alias or a stub script in your path so as to always run the current checkout's instance.
```
Comment (kraftbj) [inline docs/development-environment-cli-support.md:21]: Nice add.

File: docs/development-environment.md [hunk 2]
```diff
@@ -167,34 +167,34 @@ There are three types of builds:
...
 
 We recommend usage of [nvm](https://github.com/nvm-sh/nvm/) for managing different Node versions on the same environment.
 
-**Note:** If you have previously run the Jetpack build tasks (e.g. `yarn build`), and didn't come back to it for a long time, you can
+**Note:** If you have previously run the Jetpack build tasks (e.g. `pnpm build`), and didn't come back to it for a long time, you can
 run this command before building again. Otherwise you may experience errors on the command line while trying to build.
 
 ```sh
-yarn distclean
+pnpm distclean
```
Comment (kraftbj) [inline docs/development-environment.md:197]: This is fine for now I think. It mentions it is deprecated for me, but we can wait until the CLI clean command is ready before updating the docs.

File: package.json [hunk 3]
```diff
@@ -85,34 +85,15 @@
...
-		"**/gulp/gulp-cli/yargs/yargs-parser": "^20.0.0",
-		"**/ini": "^1.3.7",
-		"**/lodash": "4.17.21",
-		"**/node-fetch": "^2.6.1",
-		"**/node-notifier": "^8.0.1",
-		"**/tar-stream/bl": "^4.0.0",
-		"**/terser-webpack-plugin/serialize-javascript": "^5.0.0"
-	},
 	"dependencies": {
+		"chalk": "4.1.1",
```
Comment (kraftbj) [inline package.json:89]: Only a couple of deps we hadn't declared. That's not too bad, methinks.

File: projects/plugins/jetpack/_inc/lib/admin-pages/class.jetpack-react-page.php [hunk 4]
```diff
@@ -174,7 +174,7 @@ function page_render() {
 			// If we still have nothing, display an error
 			echo '<p>';
 			esc_html_e( 'Error fetching static.html. Try running: ', 'jetpack' );
-			echo '<code>yarn distclean && yarn build</code>';
+			echo '<code>pnpm run distclean && pnpm run build</code>';
```
Comment (kraftbj) [inline projects/plugins/jetpack/_inc/lib/admin-pages/class.jetpack-react-page.php:177]: While we're here, should update this to the cli command.

```suggestion
			echo '<code>pnpm run distclean && pnpx jetpack build plugins/jetpack</code>';
```

**Review by kraftbj — DISMISSED** [review]
Still want to do some more testing and didn't quite get through all of it yet, but here's a few minor things so far.

**Comment by kraftbj** [comment]
Upon switching to the branch:
2021-05-20 at 10 14 AM

Following the directions to delete `node_modules` resolves it (with the usual error about husky-run when the repo isn't installed), but should be ready to explain to folks what this means post-merge. I'm looking at making a script to help with the transition; I'll PR that separately when I get it ready-enough.

File: docs/development-environment.md (hunk 2 above)
Comment (anomiex) [reply docs/development-environment.md:197]: Yeah, as I noted in the commit message on eb625dfdc7 updating all the outdated docs seems out of scope for this PR.

File: package.json (hunk 3 above)
Comment (anomiex) [reply package.json:89]: I didn't try running all the tools manually, but we don't have too many that are written in js.

File: package.json [hunk 5]
```diff
@@ -13,65 +13,65 @@
...
+		"docker:ngrok-down": "pnpm run docker:ngrok down",
+		"docker:ngrok-up": "pnpm run docker:ngrok up",
+		"docker:phpunit": "pnpm run docker:compose -- exec wordpress phpunit --configuration=/var/www/html/wp-content/plugins/jetpack/phpunit.xml.dist",
+		"docker:phpunit:multisite": "pnpm run docker:compose -- exec wordpress phpunit --configuration=/var/www/html/wp-content/plugins/jetpack/tests/php.multisite.xml",
+		"docker:run-extras": "pnpm run docker:compose -- exec wordpress bash -c \"chmod +x /var/scripts/run-extras.sh && . /var/scripts/run-extras.sh\"",
+		"docker:sh": "pnpm run docker:compose exec wordpress bash",
+		"docker:stop": "pnpm run docker:compose stop",
+		"docker:tail": "pnpm run docker:compose exec wordpress bash -c \"/var/scripts/tail.sh\"",
+		"docker:uninstall": "pnpm run docker:compose -- exec wordpress bash -c \"/var/scripts/uninstall.sh\"",
+		"docker:up": "pnpm run docker:compose up",
```
Comment (jeherve) [inline package.json:61]: Should it be `pnpm run docker:compose -- up` here?

```suggestion
		"docker:up": "pnpm run docker:compose -- up",
```

I used to run `yarn docker:up -d` to have my container running in the background, but this now only works if I do `pnpm run docker:compose -- up -d`.

Comment (anomiex) [reply package.json:61]: Hmm. Yeah, we need the `--` in a lot of places to make sure the calls go through correctly. More than just here.

Comment (anomiex) [reply package.json]: 9527526

**Review by jeherve — COMMENTED** [review]
Builds seem to work well for me. I just cannot figure out how to run Docker in the background anymore. With yarn, I used to do `yarn docker:up -d && yarn docker:jt-up` to start working. I wouldn't see the logs from Docker that way.

`pnpm docker:up -d` doesn't seem to work that way for some reason.

File: .gitignore [hunk 6]
```diff
@@ -22,7 +22,6 @@ npm-debug.log
 *.unison.tmp
 .eslintcache
 .eslintignore-required
-yarn-error.log
 .phpcs.xml
```
Comment (anomiex) [inline .gitignore:25]: Sure.
```suggestion
.phpcs.xml
.pnpm-debug.log
```

**Review by anomiex — COMMENTED** [review]
It interprets the `-d` as a parameter to `pnpm` rather than passing it on to the script. You'd need to do `pnpm docker:up -- -d`. Or better, `jetpack docker up -d`.

**Comment by zinigor** [comment]
Thanks for working on this update! I have been testing it in various scenarios, and I get a weird problem when trying to `npx jetpack install --all`:

```
❯ Installing plugins/backup
    ✔ Installing Composer Dependencies
    ✖ Installing Pnpm Dependencies
      →  ERROR  ENOENT: no such file or directory, chmod '/home/zinigor/workspace/jetpack/projects/github-
…

  ----- SNIP -----

Error: Command failed with exit code 1: pnpm install
Scope: all 12 workspace projects
 ERROR  ENOENT: no such file or directory, chmod '/home/zinigor/workspace/jetpack/projects/github-actions/required-review/node_modules/.bin/ncc'


    at makeError (/home/zinigor/workspace/jetpack/node_modules/.pnpm/execa@5.0.0/node_modules/execa/lib/error.js:59:11)
    at handlePromise (/home/zinigor/workspace/jetpack/node_modules/.pnpm/execa@5.0.0/node_modules/execa/index.js:114:26)
    at processTicksAndRejections (internal/process/task_queues.js:95:5) {
  shortMessage: 'Command failed with exit code 1: pnpm install',
  command: 'pnpm install',
  exitCode: 1,
  signal: undefined,
  signalDescription: undefined,
  stdout: 'Scope: all 12 workspace projects\n' +
    " ERROR  ENOENT: no such file or directory, chmod '/home/zinigor/workspace/jetpack/projects/github-actions/required-review/node_modules/.bin/ncc'\n" +
    '\n',
  stderr: '',
  failed: true,
  timedOut: false,
  isCanceled: false,
  killed: false,
  context: [Object: null prototype] {}
}
```

Strangely enough, it doesn't happen when I try to `npx jetpack install` and manually choose `projects/backup`.

**Comment by jeherve** [comment]
> It interprets the `-d` as a parameter to `pnpm` rather than passing it on to the script. You'd need to do `pnpm docker:up -- -d`. Or better, `jetpack docker up -d`.

That worked, thank you!

For the Jurassic Tube part, I opened 2-gh-jurassic-tube

**Comment by anomiex** [comment]
@zinigor Looks like pnpm can race with itself if we run multiple installs for the same workspace at once. At least that's better than the situation with yarn, where it would race with itself even when running in different workspaces. a7a72b38 should avoid the multiple parallel runs.

P.S. You should probably use `pnpx` rather than `npx` there, but that doesn't seem to have anything to do with the problem you saw.

**Comment by zinigor** [comment]
Thanks Brad! The last commit has fixed the installation for me!

**Review by sdixon194 — APPROVED** [review]

**Comment by anomiex** [comment]
r226560-wpcom

**Pull request merged**